### ISSUE TITLE
Separate Emulated Operating System from Minetest for Enhanced Code Reusability and Testing

### ISSUE THREAD

**Issue by JamesClarke7283** [Hacknique/modular_computers#13]
Title: Separate Emulated Operating System from Minetest for Enhanced Code Reusability and Testing

- **Code Reusability**:
  - Propose segregating the emulated OS into a distinct Lua package to maximize code reuse.

- **User Installation**:
  - Acknowledge that users would need to install this separate package to use the mod.

- **Project Robustness**:
  - Anticipate a more robust project by isolating technical challenges related to the OS from the game mechanics.

- **Testing and Compliance**:
  - Envision easier testing routines to ensure POSIX compliance of the OS.

- **Focus**:
  - Allows the team to concentrate more on addressing OS technical hurdles.

**Comment by y5nw** [comment]
>     * **Code Reusability**:
>       
>       * Propose segregating the emulated OS into a distinct Lua package to maximize code reuse.

* _What_ code should be "reused"?
* How will "segregating the emulated OS into a distinct Lua package" maximize code reuse?
* Assuming this is done to allow the use of "different" OSes, where would the differences be? (Note that part of the kernel space, e.g. filesystems, I/O, "executables", are handled at a lower, i.e. "simulator" level)

>     * **Project Robustness**:
>       
>       * Anticipate a more robust project by isolating technical challenges related to the OS from the game mechanics.
>
>     * **Testing and Compliance**:
>       
>       * Envision easier testing routines to ensure POSIX compliance of the OS.
> 
>     * **Focus**:
>       
>       * Allows the team to concentrate more on addressing OS technical hurdles.

How will any of these by achieved by separating things into different packages?

While I do support moving less related things into different mods (and creating a modpack for these mods), this only allows users more choice, but using this to solve the issues mentioned above would be an XY problem, as the code is already organized hierarchically (making it theoretically possible to split the mods into multiple mods without much significant change) and the aspects you mentioned are more relevant to the actual quality of the code.

IMO the solution to make the project robust is to improve the code itself to allow various APIs to form and become stale, I am disappointed to see this aspect absent in the issue.

**Comment by JamesClarke7283** [comment]
> > ```
> > * **Code Reusability**:
> >   
> >   * Propose segregating the emulated OS into a distinct Lua package to maximize code reuse.
> > ```
> 
>     * _What_ code should be "reused"?
> 
>     * How will "segregating the emulated OS into a distinct Lua package" maximize code reuse?
> 
>     * Assuming this is done to allow the use of "different" OSes, where would the differences be? (Note that part of the kernel space, e.g. filesystems, I/O, "executables", are handled at a lower, i.e. "simulator" level)
> 
> 
> > ```
> > * **Project Robustness**:
> >   
> >   * Anticipate a more robust project by isolating technical challenges related to the OS from the game mechanics.
> > 
> > * **Testing and Compliance**:
> >   
> >   * Envision easier testing routines to ensure POSIX compliance of the OS.
> > 
> > * **Focus**:
> >   
> >   * Allows the team to concentrate more on addressing OS technical hurdles.
> > ```
> 
> How will any of these by achieved by separating things into different packages?
> 
> While I do support moving less related things into different mods (and creating a modpack for these mods), this only allows users more choice, but using this to solve the issues mentioned above would be an XY problem, as the code is already organized hierarchically (making it theoretically possible to split the mods into multiple mods without much significant change) and the aspects you mentioned are more relevant to the actual quality of the code.
> 
> IMO the solution to make the project robust is to improve the code itself to allow various APIs to form and become stale, I am disappointed to see this aspect absent in the issue.

Firstly, the code that would be reused is the operating system code.

I am not talking about making a modpack, but rather making the Operating System code seperate from the game.

I.E Making a luarocks package.

It will make the code more versatile because it can now be used in various different environments, not just limited to minetest.

I think this is the main advantage.

Ive made a repository for this OS: [here](https://github.com/Hacknique/Luix)

**Comment by nitrogenez** [comment]
> > > ```
> > > * **Code Reusability**:
> > >   
> > >   * Propose segregating the emulated OS into a distinct Lua package to maximize code reuse.
> > > ```
> > 
> >     * _What_ code should be "reused"?
> > 
> >     * How will "segregating the emulated OS into a distinct Lua package" maximize code reuse?
> > 
> >     * Assuming this is done to allow the use of "different" OSes, where would the differences be? (Note that part of the kernel space, e.g. filesystems, I/O, "executables", are handled at a lower, i.e. "simulator" level)
> > 
> > 
> > > ```
> > > * **Project Robustness**:
> > >   
> > >   * Anticipate a more robust project by isolating technical challenges related to the OS from the game mechanics.
> > > 
> > > * **Testing and Compliance**:
> > >   
> > >   * Envision easier testing routines to ensure POSIX compliance of the OS.
> > > 
> > > * **Focus**:
> > >   
> > >   * Allows the team to concentrate more on addressing OS technical hurdles.
> > > ```
> > 
> > How will any of these by achieved by separating things into different packages?
> > 
> > While I do support moving less related things into different mods (and creating a modpack for these mods), this only allows users more choice, but using this to solve the issues mentioned above would be an XY problem, as the code is already organized hierarchically (making it theoretically possible to split the mods into multiple mods without much significant change) and the aspects you mentioned are more relevant to the actual quality of the code.
> > 
> > IMO the solution to make the project robust is to improve the code itself to allow various APIs to form and become stale, I am disappointed to see this aspect absent in the issue.
> 
> Firstly, the code that would be reused is the operating system code.
> 
> I am not talking about making a modpack, but rather making the Operating System code seperate from the game.
> 
> I.E Making a luarocks package.
> 
> It will make the code more versatile because it can now be used in various different environments, not just limited to minetest.
> 
> I think this is the main advantage.
> 
> Ive made a repository for this OS: [here](https://github.com/Hacknique/Luix)

Lua outside of MT is very different and should be architectured differently, for example `dofile()` is not a really good practice in real Lua packages, as that is an analog of `require`, which is morr suitable.

I personally think that nothing than code maintenance difficulty has changed.

**Comment by y5nw** [comment]
> > > ```
> > > * **Code Reusability**:
> > >   
> > >   * Propose segregating the emulated OS into a distinct Lua package to maximize code reuse.
> > > ```
> > 
> > 
> > ```
> > * _What_ code should be "reused"?
> > 
> > * How will "segregating the emulated OS into a distinct Lua package" maximize code reuse?
> > 
> > * Assuming this is done to allow the use of "different" OSes, where would the differences be? (Note that part of the kernel space, e.g. filesystems, I/O, "executables", are handled at a lower, i.e. "simulator" level)
> > ```
> > 
> > 
> > > ```
> > > * **Project Robustness**:
> > >   
> > >   * Anticipate a more robust project by isolating technical challenges related to the OS from the game mechanics.
> > > 
> > > * **Testing and Compliance**:
> > >   
> > >   * Envision easier testing routines to ensure POSIX compliance of the OS.
> > > 
> > > * **Focus**:
> > >   
> > >   * Allows the team to concentrate more on addressing OS technical hurdles.
> > > ```
> > 
> > 
> > How will any of these by achieved by separating things into different packages?
> > While I do support moving less related things into different mods (and creating a modpack for these mods), this only allows users more choice, but using this to solve the issues mentioned above would be an XY problem, as the code is already organized hierarchically (making it theoretically possible to split the mods into multiple mods without much significant change) and the aspects you mentioned are more relevant to the actual quality of the code.
> > IMO the solution to make the project robust is to improve the code itself to allow various APIs to form and become stale, I am disappointed to see this aspect absent in the issue.
> 
> Firstly, the code that would be reused is the operating system code.
> 
> I am not talking about making a modpack, but rather making the Operating System code seperate from the game.
> 
> I.E Making a luarocks package.
> 
> It will make the code more versatile because it can now be used in various different environments, not just limited to minetest.
> 
> I think this is the main advantage.

That would be a :-1: from me.

Instead of repeating nitrogenez' message, I would like to point out further implications of this separation.

If you are going to integrate the OS elsewhere (e.g. ncurses and a regular Lua environment), you would need a lot of code in addition to make the OS able to adapt to these different platforms and, more notably, their different capabilities. For example, porting this to run on an ncurses-based platform would allow you to create an input system with extended capabilities (getting individual key input, for example). However, this "enhanced code reusability" would effectively waste this potential at the same time, as the OS (and, by extension, programs written for this OS) is also expected to work in MT. Alternatively, you can have people write code that depends on the capabilities or behave differently depending on the capability of the platform. The OS would also need to face this choice, and, in the latter case, the platform-dependent part of the codebase would likely make a separate project more sensible.

TL;DR: You would basically increase code reusability at the expense of either wasting the potential of more capable platforms or requiring developers (including that of the OS, for that matter) to "write once, debug everywhere".

**Comment by JamesClarke7283** [comment]
Okay i have heard your concerns, i won't be making a separate package for this mod for the OS. At least, as part of this project.
I will keep it packaged inside the mod.

I am closing this issue, i will reopen it if there is any more interest.
